certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,556 @@
+        ( f.selector == sig:setPendingRoot(address).selector ||
+        f.selector == sig:claimRoot().selector ),
+        "Pending root changed by a function other than setPendingRoot or claimRoot.";
+    // if the function changed the pending root, the sender was pending root or execution helper.
+    assert pendingRootBefore != pendingRootAfter =>
+        e.msg.sender == pendingRootBefore ||
+        e.msg.sender == executionHelper,
+        "Pending root changed by somebody, who was not root.";
+    assert (pendingRootBefore == 0 && pendingRootAfter != 0) => e.msg.sender == executionHelper,
+        "pendingRoot can only be changed by the executer";
+}
+
+
+// STATUS - verified
+// ScheduledExecution that has already been cancelled or executed
+// cannot be canceled again.
+// https://prover.certora.com/output/40577/0a8cadb54810435c859098750efa8ee0/?anonymousKey=58e33df6c4063dd4882ec3f40d8b86d816951102
+rule scheduledExecutionCanBeCancelledOnlyOnce(env e, uint256 index) {
+    bool canceled_before = getSchedExeCancelled(index);
+    bool executed_before = getSchedExeExecuted(index);
+
+    cancel@withrevert(e, index);
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// ScheduledExecution that has already been cancelled or executed
+// cannot be executed again.
+// https://prover.certora.com/output/40577/0a8cadb54810435c859098750efa8ee0/?anonymousKey=58e33df6c4063dd4882ec3f40d8b86d816951102
+rule scheduledExecutionCanBeExecutedOnlyOnce(env e, uint256 index) {
+    bool canceled_before = getSchedExeCancelled(index);
+    bool executed_before = getSchedExeExecuted(index);
+
+    execute@withrevert(e, index);
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// Scheduled execution can only be executed by those supplied as executors during scheduling
+// https://prover.certora.com/output/40577/f84672d68a6d42498d6899d27b9288a1/?anonymousKey=cf4c6f18c3a8fad6fc659a983a8ed595775a061a
+rule cannotBecomeExecutorForAlreadyScheduledExecution(env e, method f) {
+    uint256 length = getSchedExeLength();
+    uint256 id;
+    address user;
+    require(length < max_uint256);
+    require(id < length);
+    bool isExecutor = isExecutor(id, user);
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    assert !isExecutor => !isExecutor(id, user),
+        "User became an executor after the scheduled execution has been scheduled.";
+}
+
+
+// STATUS - verified
+// Only execution, that has not yet been cancelled or executed can be executed
+// and only if it's executableAt is not greater than current timestamp.
+// Also, when execution happens, the caller must be executor.
+// Also, executor can execute execution, which is executable (`canExecute`)
+// https://prover.certora.com/output/40577/0b63385b7d2c4389acfcd10e8e4babfa/?anonymousKey=602c9b0b12b1e83f7625b142d5244a03f6ef6a3c
+rule whoCanExecuteAndWhatIsExecutable(env e) {
```

Comment:
Probably it's better to split it into several rules. Too overwhelming IMHO. But here decide as you want because you probably don't have time for it.

Reply:
That is an excellent point, I split the rule into `whatCanBeExecuted` and `whoCanExecute` and after I verify I will also reflect it in the report.

Reply:
Both rules are passing [here](https://vaas-stg.certora.com/output/40577/2feafb4a216e48f199389a3e47370c79/?anonymousKey=82653964f7ef8ebe6bee53655c0aacf33c8b6625) and [here](https://vaas-stg.certora.com/output/40577/6caa1ef684e448b29a7cff873d8d31ca/?anonymousKey=7e05e33d21a27b1a80b5598ad822a5061eb426ad). I also modified the report.